### PR TITLE
#3262 -  Ministry Student Application Tracker

### PR DESCRIPTION
**Pull request by dheepak-aot**
# Ministry Student Application Tracker - Part 1

## API
API used by student application tracker: `application/:applicationId/in-progress`  `application/:applicationId/progress-details`  `application/:applicationId/enrolment`  `application/:applicationId/completed` 

- [x] For all the APIs used for student application tracker, the API implementation is centralized using controller service.
- [x] Created ministry endpoints for all student application tracker APIs
- [x] Minor update to `application/:applicationId/in-progress` throw error if the application status is not `in-progress`.
- [x] Minor refactor to not repeat the security validation with student id in all places and do it only once. (Removed from places 2 and 3).
image

## E2E Tests

- [x] Added 2 tests, happy path and not found error for all the APIs.

image

image

image

image

- [x] Added `pirStatus` to application factory methods.

### CODE REVIEW

File: sources/packages/backend/apps/api/src/route-controllers/application/_tests_/application.aest.controller.getApplicationProgressDetails.e2e-spec.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    const token = await getAESTToken(AESTGroups.BusinessAdministrators);
+
+    // Act/Assert
+    await request(app.getHttpServer())
+      .get(endpoint)
+      .auth(token, BEARER_AUTH_TYPE)
+      .expect(HttpStatus.NOT_FOUND)
+      .expect({
+        statusCode: HttpStatus.NOT_FOUND,
+        message: `Application id 99999999 was not found.`,
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application/_tests_/application.aest.controller.getApplicationProgressDetails.e2e-spec.ts:103]: Use double quotes :)

File: sources/packages/backend/apps/api/src/route-controllers/application/_tests_/application.aest.controller.getCompletedApplicationDetails.e2e-spec.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    const token = await getAESTToken(AESTGroups.BusinessAdministrators);
+
+    // Act/Assert
+    await request(app.getHttpServer())
+      .get(endpoint)
+      .auth(token, BEARER_AUTH_TYPE)
+      .expect(HttpStatus.NOT_FOUND)
+      .expect({
+        statusCode: HttpStatus.NOT_FOUND,
+        message: `Application not found or not on Completed status.`,
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application/_tests_/application.aest.controller.getCompletedApplicationDetails.e2e-spec.ts:99]: Use double quotes :)

File: sources/packages/backend/apps/api/src/route-controllers/application/_tests_/application.aest.controller.getEnrolmentApplicationDetails.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    const token = await getAESTToken(AESTGroups.BusinessAdministrators);
+
+    // Act/Assert
+    await request(app.getHttpServer())
+      .get(endpoint)
+      .auth(token, BEARER_AUTH_TYPE)
+      .expect(HttpStatus.NOT_FOUND)
+      .expect({
+        statusCode: HttpStatus.NOT_FOUND,
+        message: `Application id 99999999 not found or not in relevant status to get enrolment details.`,
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application/_tests_/application.aest.controller.getEnrolmentApplicationDetails.e2e-spec.ts:77]: Double quotes

**Review by andrepestana-aot — APPROVED** [review]
Looks good to me!

File: sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts [hunk 4]
```diff
@@ -294,12 +295,20 @@ export class StudentAppealService extends RecordDataModelService<StudentAppeal>
...
-      .orderBy("studentAppeal.submittedDate", "DESC");
+      .where("application.id = :applicationId", { applicationId });
+
+    if (options?.studentId) {
+      query.andWhere("application.student.id = :studentId", {
+        studentId: options.studentId,
+      });
+    }
+
+    query.orderBy("studentAppeal.submittedDate", "DESC");
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts:306]: Minor, I believe the `orderBy` can be added in the main query after the first `where`.

**Review by andrewsignori-aot — APPROVED** [review]
Amazing work, looks good 👍

File: sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts (hunk 4 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts:306]: I will test and try to add that in upcoming PR.

**Pull request merged**